node/node_newblock.go

```diff
@@ -83,14 +84,27 @@ func (node *Node) proposeNewBlock() (*types.Block, error) {
 	node.Worker.UpdateCurrent()
 
 	// Update worker's current header and state data in preparation to propose/process new transactions
-	coinbase := node.Consensus.SelfAddress
+	var (
+		coinbase    = node.Consensus.SelfAddress
+		beneficiary = coinbase
+		err         error
+	)
 
 	// After staking, all coinbase will be the address of bls pub key
 	if node.Blockchain().Config().IsStaking(node.Worker.GetCurrentHeader().Epoch()) {
 		addr := common.Address{}
-		blsPubKeyBytes := node.Consensus.PubKey.GetAddress()
-		addr.SetBytes(blsPubKeyBytes[:])
-		coinbase = addr
+		if bytes.Compare(coinbase[:], addr[:]) == 0 { // empty SelfAddress means it's a staking validator
```

Comment:
what if coinbase not equal bls address (unless selfaddress always equal bls address in this case)?

Reply:
Here it's checking whether SelfAddress is empty. If it's empty, it means it's staking validator, and the ECDSA address needs to be looked up.